logstash-core/src/test/java/org/logstash/plugin/InputTest.java

```diff
@@ -0,0 +1,46 @@
+package org.logstash.plugin;
+
+import org.junit.Test;
+import org.logstash.Event;
+import org.logstash.TestUtil;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.function.Consumer;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+public class InputTest {
+    private Collection<Collection<Event>> batches = new LinkedList<>();
+    private int batchCount = TestUtil.random.nextInt(10) + 1;
+
+    @Test
+    public void testInput() {
+        Input i = new TestInput();
+        i.run(batches::add);
+        assertEquals(batchCount, batches.size());
+        for (Collection<Event> events : batches) {
+            assertFalse(events.isEmpty());
+        }
+    }
+
+    private class TestInput implements Input {
+        @Override
+        public void run(Consumer<Collection<Event>> consumer) {
+            for (int b = 0; b < batchCount; b++) {
+                int eventCountPerBatch = TestUtil.random.nextInt(100) + 1;
+                Collection<Event> events = new LinkedList<>();
+                for (int i = 0; i < eventCountPerBatch; i++) {
+                    Event e = new Event();
+                    e.setField("message", "hello world");
+                    e.setField("i", i);
+                    e.setField("b", b);
+                    events.add(e);
+                }
+                consumer.accept(Collections.unmodifiableCollection(events));
```

Comment:
@jordansissel this is actually a good example of why you don't want to expose a `Collection<>` API to plugins here. This makes performance really unpredictable.
If someone passes us a memory-wise cheap `ArrayList` here, performance characteristics (especially optimal on-heap queue size(s)) will be very different from a super heavy `LinkedList`. Only allowing single events to be passed here, allows us to have much more predictable (and if we do a good job, also faster) behavior here imo.

Reply:
I'm in favor of your EventIterator approach. We'll want EventIterator to have `remove` also, eventually (to replace Event.cancel). 

But what about the return type? (Where a processor is allowed to create events, such as the split filter, perhaps we should make this a special case instead of the common case?)

For the Processor return value, we'll have the same problem if someone passes us an EventIterator that is backed by a LinkedList.iterator(), right?

This is specific to the Processor's return type, should it return a collection? a list? a set? a custom type? If not return type, should a Processor become a BiConsumer that has this signature:

```
void process(EventIterator events, EventProducer newEvents) {
...
}
```

where `newEvents` has a small interface to allow processors to create events (like, `newEvents.add(event)` and Logstash takes care of the actual implementation)

Thoughts?

Reply:
@jordansissel 

>We'll want EventIterator to have remove also, eventually (to replace Event.cancel).

I think I'd just make it `EventIterator  process(EventIterator events)` here and handle `remove` by just callling `next` without pushing an `Event` to the returned `EventIterator`.

Same for adding events. I'd just make this an implementation detail of how a `Filter` implements the `EventIterator` it returns.
We can always offer something like `SimpleFilter` that wraps this process into say `Collection process(Collection<Event> events)` to make it easier to implement things for third party developers (and we need that for existing Ruby stuff anyhow), but in general a `Filter` should simply be a mapping `EventIterator -> EventIterator` imo.